Add build event for volume streaming

andy-paine opened the pull request:
## Changes proposed by this PR

Often build steps appear to "hang" in the UI after a worker is selected
which is commonly due to volumes being streamed. As these volumes can be
quite large, this can take a significant amount of time that would be
helpful to tell the user about.

This adds a "streaming-volume" event to the build logs which outputs the
"input name" (directory that it will be mounted at) and the source and
destination worker names to help users understand when jobs are waiting
for volumes to be streamed.

The current worker structure did not really have a particularly good
place to save this event so I kept the "delegate" pattern and passed
that down into the atc/worker code. Due to cyclic dependencies this has
to be in the atc/runtime package but I think that is fine anyway as the
runtime.BuildStepDelegate only specifies the delegate methods needed at
runtime etc.

closes #6575

* [x] done
* [ ] todo

## Notes to reviewer
The delegate being passed all the way down seems like an elegant-ish hack but open to other suggestions

## Release Note
* Build logs will now contain new events when a volume is being streamed to a worker
![Screenshot 2022-02-02 at 11 39 01](https://user-images.githubusercontent.com/14118619/152146829-d058dcbd-6db7-40ec-a0b0-c53306d119de.png)

Commit:
atc: Add build event for volume streaming

Often build steps appear to "hang" in the UI after a worker is selected
which is commonly due to volumes being streamed. As these volumes can be
quite large, this can take a significant amount of time that would be
helpful to tell the user about.

This adds a "streaming-volume" event to the build logs which outputs the
"input name" (directory that it will be mounted at) and the source and
destination worker names to help users understand when jobs are waiting
for volumes to be streamed.

The current worker structure did not really have a particularly good
place to save this event so I kept the "delegate" pattern and passed
that down into the atc/worker code. Due to cyclic dependencies this has
to be in the atc/runtime package but I think that is fine anyway as the
runtime.BuildStepDelegate only specifies the delegate methods needed at
runtime etc.

Signed-off-by: Andy Paine <andy.paine@engineerbetter.com>

Commit:
atc: Move runtime.Artifact interface into runtime.Volume

The commit message in 36dda00 says that this was created to allow for
other Artifact sources rather than just runtime.Volumes. At the time of
writing, no other Artifact implementations exist.

If any new implementations get created then it seems reasonable that
they should have a Handle (unique ID) and a source (e.g. ATC) so the
extra methods will be natural to implement. This saves some casting in
places where it currently is _always_ a runtime.Volume

Rather than using the var _ Artifact = (Volume)(nil) static check, just
embed Artifact in Volume. There are some unfortunate places such as
streamed resource cache stuff where we still need to cast to an explicit
type but that is a larger change that may not fit as well into the
abstraction as Source and Handle seem to so I've left that

Signed-off-by: Andy Paine <andy.paine@engineerbetter.com>

andy-paine commented on atc/worker/gardenruntime/worker.go:
```diff
@@ -43,6 +44,7 @@ type DB struct {
 	WorkerTaskCacheFactory        db.WorkerTaskCacheFactory
 	ResourceCacheFactory          db.ResourceCacheFactory
 	WorkerBaseResourceTypeFactory db.WorkerBaseResourceTypeFactory
+	BuildFactory                  db.BuildFactory
```
This is a mistake

andy-paine commented on atc/runtime/types.go:
```diff
@@ -276,26 +281,30 @@ type ContainerLimits struct {
 type Volume interface {
-	// Handle gives the globally unique ID of the Volume.
-	Handle() string
+	Artifact
```
Currently in lots of places that we use `runtime.Artifact` we end up casting it to `runtime.Volume` to be able to call functions to get the handle and source worker. The `runtime.Artifact` endpoint is designed to be a future abstraction that only `gardenruntime.Volume` currently implements and it seems reasonable that any future implementation should also be able to say what its unique identifier is `Handle()` and where it came from `Source()`. By moving these functions up to the `runtime.Artifact` interface, we avoid having to cast it to `runtime.Volume` in places like this https://github.com/concourse/concourse/pull/8031/files#diff-2034974614561f2b31bff2d4e9572aa96a162533d58b49a6cc06cfd74625d905R522

xtremerui commented on atc/syslog/drainer.go:
```diff
@@ -173,6 +173,16 @@ func (d *drainer) sendEvent(logger lager.Logger, build db.Build, syslog *Syslog,
+		}
+		ts = time.Unix(streamingVolumeEvent.Time, 0)
+		tag = build.SyslogTag(streamingVolumeEvent.Origin.ID)
+		message = fmt.Sprintf("streaming volume %s from %s to %s", streamingVolumeEvent.Volume, streamingVolumeEvent.SourceWorker, streamingVolumeEvent.DestWorker)
```
```suggestion
		message = fmt.Sprintf("streaming volume %s from worker %s to %s", streamingVolumeEvent.Volume, streamingVolumeEvent.SourceWorker, streamingVolumeEvent.DestWorker)
```
I like the way `streaming volume xxx` so i know `xxx` is a volume. Could we also do that for worker?

Another question is the `to %s` part, is there a case that destination worker is different to the worker in `selected worker` event just above? If not, I feel `streaming volume xxx from worker xxx` might be good enough.

Commit:
Reword streaming volume message

Be explicit that the source is a worker in messages (in web and fly
stream) and omit the destination since the `selected-worker` event shows
that and this would just be duplicate info.

I've left the DestWorker on the actual event since it could be useful as
part of debugging or an audit trail. Omitting it in output makes sense
as all the events around it give context but by leaving the DestWorker
in, this means you don't need to "piece together" what happened before
the `streaming-volume` event and can just look at it in isolation

Signed-off-by: Andy Paine <andy.paine@engineerbetter.com>

andy-paine commented:
Any idea why this PR's pipeline is archived - the checks won't run 😞 ? https://ci.concourse-ci.org/teams/contributor/pipelines/pr?vars.number=8031

wanderanimrod commented:
@andy-paine , do we know the size of the volume being streamed at the location we are logging this? It would be very nice to indicate the size in human-readable units so that users have more data while optimizing their pipelines. Just asking before I open a new issue for this. If it is too hard, I'd rather not open an issue that will be stale for long. Thanks.
